src/patterns.jl

```diff
@@ -0,0 +1,63 @@
+
+"""
+    Pattern
+
+A pattern is just a wrapper around a `Tuple{Vararg{Symbol}}` which enables searching for
+matching components and dimension paths in a Dataset.
+Special symbols `:_` and `:__` are used as wildcards, similar to `*` and `**` in glob
+pattern matching.
+"""
+struct Pattern
+    segments::Tuple{Vararg{Symbol}}
+end
+
+Pattern(segments::Symbol...) = Pattern(segments)
+Base.convert(::Type{Pattern}, x::Tuple{Vararg{Symbol}}) = Pattern(x)
+
+# Primary functionality is to support identify tuples that match the constraint
+# NOTE: I'm not entirely sure if we should overload `in` for this algorithm
```

Comment:
how about `match`? Base uses it for regex pattern matching but it fits semantically with what we're doing too:

>A pattern...enables searching for _matching_ components and dimension paths

Reply:
The problem is that `match` returns a `RegexMatch` object, and we just want to return a `Bool`. Similarly, `occursin` seems to be restricted to strings and our args would be backwards from the regex call. I feel like `in` or `ismatch` are probably the most appropriate here?

Reply:
`in` doesn't quite fit though. Do we have to overload Base? Can't we just define `ismatch` or `matches` within this module?